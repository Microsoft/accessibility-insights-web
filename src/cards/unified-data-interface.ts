// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

// this is similar to `TestEngine` interface from axe-core
interface ScanEngine {
    name: string;
    version: string;
}

interface OSData {
    name: string;
    osVersion?: string;
    userAgent?: string;
}

interface ViewPortData {
    platformHeight?: number;
    platformWidth?: number;
}

interface PlatformData {
    osInfo: OSData;
    viewPortInfo: ViewPortData;
}

interface ToolData {
    scanEngine: ScanEngine;
}

interface ScanMetaData {}

interface ScanResults {
    results: UnifiedResults;
    platformInfo: PlatformData;
    toolInfo: ToolData;
    scanMetaData: ScanMetaData;
}

interface UnifiedResults {
    passes: UnifiedResultInstance[];
    violations: UnifiedResultInstance[];
    incomplete: UnifiedResultInstance[];
}

interface UnifiedRuleResult {
    id: string;
    ruleDescription: string;
}

interface InstancePropertyBag<T> {
    [property: string]: T;
}

type StoredInstancePropertyBag = InstancePropertyBag<any>;

interface UnifiedResultInstance {
    id: string;
    evaluation: StoredInstancePropertyBag;
    identifiers: StoredInstancePropertyBag;
    descriptors: StoredInstancePropertyBag;
    resolution: StoredInstancePropertyBag;
}
