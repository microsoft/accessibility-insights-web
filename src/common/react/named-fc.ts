// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export type ReactFCWithDisplayName<P = {}> = React.FC<React.PropsWithChildren<P>> & {
    displayName: string;
};

export function NamedFC<P = {}>(
    displayName: string,
    component: React.FC<React.PropsWithChildren<P>>,
): ReactFCWithDisplayName<P> {
    component.displayName = displayName;

    return component as ReactFCWithDisplayName<P>;
}
