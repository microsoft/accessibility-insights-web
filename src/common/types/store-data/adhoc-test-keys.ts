// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
export const enum AdHocTestkeys {
    Color = 'color',
    Headings = 'headings',
    Issues = 'issues',
    Landmarks = 'landmarks',
    TabStops = 'tabStops',
    NeedsReview = 'needsReview',
}
