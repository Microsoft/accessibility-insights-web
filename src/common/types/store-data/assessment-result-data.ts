// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DictionaryStringTo } from '../../../types/common-types';
import { Tab } from '../../itab';
import { ManualTestStatus, ManualTestStatusData } from '../manual-test-status';
import { VisualizationType } from '../visualization-type';

export type TestStepInstance = UserCapturedInstance & GeneratedAssessmentInstance;

export type PersistedTabInfo = Tab & {
    appRefreshed: boolean;
};

export interface AssessmentStoreData {
    persistedTabInfo: PersistedTabInfo;
    assessments: {
        [key: string]: AssessmentData;
    };
    assessmentNavState: AssessmentNavState;
}

export type InstanceIdToInstanceDataMap = DictionaryStringTo<GeneratedAssessmentInstance>;
export type RequirementIdToResultMap = DictionaryStringTo<ManualTestStepResult>;

export interface AssessmentData {
    fullAxeResultsMap: any;
    generatedAssessmentInstancesMap?: InstanceIdToInstanceDataMap;
    manualTestStepResultMap?: RequirementIdToResultMap;
    testStepStatus: ManualTestStatusData;
}

export interface ManualTestStepResult {
    status: ManualTestStatus;
    id: string;
    instances: UserCapturedInstance[];
}

export interface UserCapturedInstance {
    id: string;
    description: string;
    html?: string;
    selector?: string;
}

export interface GeneratedAssessmentInstance<T = {}, K = {}> {
    target: string[];
    html: string;
    testStepResults: IAssessmentResultType<K>;
    propertyBag?: T;
}

export interface TestStepResult {
    id: string;
    status: ManualTestStatus;
    isCapturedByUser: boolean;
    failureSummary: string;
    isVisualizationEnabled: boolean;
    isVisible: boolean;
    originalStatus?: ManualTestStatus;
}

export interface AssessmentNavState {
    selectedTestStep: string;
    selectedTestType: VisualizationType;
}

// tslint:disable-next-line:interface-name
export interface IHeadingsAssessmentProperties {
    headingLevel: string;
    headingText: string;
}

// tslint:disable-next-line:interface-name
export interface IFrameAssessmentProperties {
    frameType: string;
    frameTitle?: string;
}

// tslint:disable-next-line:interface-name
export interface ILandmarksAssessmentProperties {
    role: string;
    label: string;
}

// tslint:disable-next-line:interface-name
export type IAssessmentInstancesMap<T = {}, K = {}> = DictionaryStringTo<GeneratedAssessmentInstance<T, K>>;
// tslint:disable-next-line:interface-name
export type IAssessmentResultType<K> = { [testStepName in keyof K]: TestStepResult };
