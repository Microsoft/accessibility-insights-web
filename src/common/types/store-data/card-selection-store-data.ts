// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export interface RuleExpandCollapseData {
    isExpanded: boolean;
    cards: { [resultInstanceUid: string]: boolean };
}

export interface CardSelectionStoreData {
    rules: { [ruleId: string]: RuleExpandCollapseData };
}
