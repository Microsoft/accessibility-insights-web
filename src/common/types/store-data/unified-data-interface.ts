// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GuidanceLink } from '../../../scanner/rule-to-links-mappings';

// this is similar to `TestEngine` interface from axe-core
export interface ScanEngineProperties {
    name: string;
    version: string;
}

export interface ApplicationProperties {
    name: string;
    version: string;
}

export interface OSProperties {
    name: string;
    version: string;
    userAgent?: string;
}

export interface ViewPortProperties {
    height?: number;
    width?: number;
}

export interface PlatformData {
    osInfo: OSProperties;
    viewPortInfo: ViewPortProperties;
}

export interface TargetAppData {
    name: string;
    url?: string;
}

export interface ToolData {
    scanEngineProperties: ScanEngineProperties;
    applicationProperties: ApplicationProperties;
}

export interface UnifiedRule {
    id: string;
    description: string;
    url: string;
    guidance: GuidanceLink[];
}

export interface UnifiedScanResultStoreData {
    results: UnifiedResult[];
    rules: UnifiedRule[];
    platformInfo?: PlatformData;
    toolInfo?: ToolData;
    targetAppInfo?: TargetAppData;
}

export interface InstancePropertyBag {
    [property: string]: any;
}

export type StoredInstancePropertyBag = InstancePropertyBag;

export type UnifiedIdentifiers = {
    identifier: string;
    conciseName: string;
} & InstancePropertyBag;

export type UnifiedDescriptors = {
    snippet?: string;
} & InstancePropertyBag;

export type UnifiedResolution = {
    howToFixSummary: string;
} & InstancePropertyBag;

export interface UnifiedResult {
    uid: string;
    status: InstanceResultStatus;
    ruleId: string;
    identifiers: UnifiedIdentifiers;
    descriptors: UnifiedDescriptors;
    resolution: UnifiedResolution;
}

export type InstanceResultStatus = 'pass' | 'fail' | 'unknown';

export type UnifiedRuleResultStatus = InstanceResultStatus | 'inapplicable';

export interface UnifiedRuleResult {
    id: string;
    nodes: UnifiedResult[];
    description: string;
    url: string;
    guidance: GuidanceLink[];
}

export type UnifiedStatusResults = {
    [key in UnifiedRuleResultStatus]: UnifiedRuleResult[];
};

export const AllRuleResultStatuses: UnifiedRuleResultStatus[] = ['pass', 'fail', 'unknown', 'inapplicable'];
