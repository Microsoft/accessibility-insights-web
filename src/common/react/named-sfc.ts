// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export type ReactSFCWithDisplayName<P = {}> = React.SFC<P> & { displayName: string };

export function NamedSFC<P = {}>(displayName: string, component: React.SFC<P>): ReactSFCWithDisplayName<P> {
    component.displayName = displayName;

    return component as ReactSFCWithDisplayName<P>;
}
