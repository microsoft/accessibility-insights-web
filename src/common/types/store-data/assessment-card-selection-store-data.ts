// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { CardSelectionStoreData } from 'common/types/store-data/card-selection-store-data';

export interface AssessmentCardSelectionStoreData {
    [testKey: string]: CardSelectionStoreData;
}
