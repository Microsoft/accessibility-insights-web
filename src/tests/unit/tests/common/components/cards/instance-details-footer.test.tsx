// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import {
    allCardInteractionsSupported,
    noCardInteractionsSupported,
    onlyUserConfigAgnosticCardInteractionsSupported,
} from 'common/components/cards/card-interaction-support';
import {
    HighlightState,
    InstanceDetailsFooter,
    InstanceDetailsFooterDeps,
    InstanceDetailsFooterProps,
} from 'common/components/cards/instance-details-footer';
import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { shallow } from 'enzyme';
import * as React from 'react';

import { exampleUnifiedResult } from './sample-view-model-data';

describe('InstanceDetailsFooter', () => {
    let resultStub: UnifiedResult;
    let props: InstanceDetailsFooterProps;
    let deps: InstanceDetailsFooterDeps;

    beforeEach(() => {
        resultStub = exampleUnifiedResult;
        deps = {
            cardInteractionSupport: allCardInteractionsSupported,
        } as InstanceDetailsFooterDeps;
        props = {
            deps,
            result: resultStub,
            highlightState: 'hidden',
        } as InstanceDetailsFooterProps;
    });

    it('renders as null when no card interactions are supported', () => {
        deps.cardInteractionSupport = noCardInteractionsSupported;
        const testSubject = shallow(<InstanceDetailsFooter {...props} />);

        expect(testSubject.getElement()).toBeNull();
    });

    it('renders per snapshot when all card interactions are supported (ie, web)', () => {
        deps.cardInteractionSupport = allCardInteractionsSupported;
        const testSubject = shallow(<InstanceDetailsFooter {...props} />);

        expect(testSubject.getElement()).toMatchSnapshot();
    });

    it('renders per snapshot when only UserConfig-agnostic card interactions are supported (ie, electron)', () => {
        deps.cardInteractionSupport = onlyUserConfigAgnosticCardInteractionsSupported;
        const testSubject = shallow(<InstanceDetailsFooter {...props} />);

        expect(testSubject.getElement()).toMatchSnapshot();
    });

    const allHighlightStates: HighlightState[] = ['visible', 'hidden', 'unavailable'];

    it.each(allHighlightStates)('renders per snapshot with highlightState="%s"', (highlightState: HighlightState) => {
        props.highlightState = highlightState;
        const testSubject = shallow(<InstanceDetailsFooter {...props} />);

        expect(testSubject.getElement()).toMatchSnapshot();
    });
});
