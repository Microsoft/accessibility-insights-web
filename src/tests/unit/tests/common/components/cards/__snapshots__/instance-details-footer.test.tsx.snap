// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstanceDetailsFooter renders per snapshot when all card interactions are supported (ie, web) 1`] = `
<div
  className="foot"
>
  <div
    className="highlightDiv"
  >
    <StyledIconBase
      ariaLabel="HighLight hidden"
      iconName="redEye"
    />
    <StyledLabelBase>
      HighLight hidden
    </StyledLabelBase>
  </div>
  <CardKebabMenuButton
    deps={
      Object {
        "cardInteractionSupport": Object {
          "supportsCopyFailureDetails": true,
          "supportsHighlighting": true,
          "supportsIssueFiling": true,
        },
      }
    }
    issueDetailsData={
      Object {
        "element": Object {
          "conciseName": "conciseName",
          "identifier": "identifier",
        },
        "howToFixSummary": "howToFixSummary",
        "rule": Object {
          "description": "description",
          "guidance": Array [
            Object {
              "href": "www.test.com",
              "tags": Array [
                Object {
                  "displayText": "New for WCAG 2.1",
                  "id": "WCAG_2_1",
                },
              ],
              "text": "text",
            },
          ],
          "id": "id",
          "url": "url",
        },
        "snippet": "snippet",
        "targetApp": Object {
          "name": "name",
          "url": "url",
        },
      }
    }
  />
</div>
`;

exports[`InstanceDetailsFooter renders per snapshot when only UserConfig-agnostic card interactions are supported (ie, electron) 1`] = `
<div
  className="foot"
>
  <div
    className="highlightDiv"
  >
    <StyledIconBase
      ariaLabel="HighLight hidden"
      iconName="redEye"
    />
    <StyledLabelBase>
      HighLight hidden
    </StyledLabelBase>
  </div>
  <CardKebabMenuButton
    deps={
      Object {
        "cardInteractionSupport": Object {
          "supportsCopyFailureDetails": true,
          "supportsHighlighting": true,
          "supportsIssueFiling": false,
        },
      }
    }
    issueDetailsData={
      Object {
        "element": Object {
          "conciseName": "conciseName",
          "identifier": "identifier",
        },
        "howToFixSummary": "howToFixSummary",
        "rule": Object {
          "description": "description",
          "guidance": Array [
            Object {
              "href": "www.test.com",
              "tags": Array [
                Object {
                  "displayText": "New for WCAG 2.1",
                  "id": "WCAG_2_1",
                },
              ],
              "text": "text",
            },
          ],
          "id": "id",
          "url": "url",
        },
        "snippet": "snippet",
        "targetApp": Object {
          "name": "name",
          "url": "url",
        },
      }
    }
  />
</div>
`;

exports[`InstanceDetailsFooter renders per snapshot with highlightState="hidden" 1`] = `
<div
  className="foot"
>
  <div
    className="highlightDiv"
  >
    <StyledIconBase
      ariaLabel="HighLight hidden"
      iconName="redEye"
    />
    <StyledLabelBase>
      HighLight hidden
    </StyledLabelBase>
  </div>
  <CardKebabMenuButton
    deps={
      Object {
        "cardInteractionSupport": Object {
          "supportsCopyFailureDetails": true,
          "supportsHighlighting": true,
          "supportsIssueFiling": true,
        },
      }
    }
    issueDetailsData={
      Object {
        "element": Object {
          "conciseName": "conciseName",
          "identifier": "identifier",
        },
        "howToFixSummary": "howToFixSummary",
        "rule": Object {
          "description": "description",
          "guidance": Array [
            Object {
              "href": "www.test.com",
              "tags": Array [
                Object {
                  "displayText": "New for WCAG 2.1",
                  "id": "WCAG_2_1",
                },
              ],
              "text": "text",
            },
          ],
          "id": "id",
          "url": "url",
        },
        "snippet": "snippet",
        "targetApp": Object {
          "name": "name",
          "url": "url",
        },
      }
    }
  />
</div>
`;

exports[`InstanceDetailsFooter renders per snapshot with highlightState="unavailable" 1`] = `
<div
  className="foot"
>
  <div
    className="highlightDiv"
  >
    <StyledIconBase
      ariaLabel="HighLight unavailable"
      iconName="redEye"
    />
    <StyledLabelBase>
      HighLight unavailable
    </StyledLabelBase>
  </div>
  <CardKebabMenuButton
    deps={
      Object {
        "cardInteractionSupport": Object {
          "supportsCopyFailureDetails": true,
          "supportsHighlighting": true,
          "supportsIssueFiling": true,
        },
      }
    }
    issueDetailsData={
      Object {
        "element": Object {
          "conciseName": "conciseName",
          "identifier": "identifier",
        },
        "howToFixSummary": "howToFixSummary",
        "rule": Object {
          "description": "description",
          "guidance": Array [
            Object {
              "href": "www.test.com",
              "tags": Array [
                Object {
                  "displayText": "New for WCAG 2.1",
                  "id": "WCAG_2_1",
                },
              ],
              "text": "text",
            },
          ],
          "id": "id",
          "url": "url",
        },
        "snippet": "snippet",
        "targetApp": Object {
          "name": "name",
          "url": "url",
        },
      }
    }
  />
</div>
`;

exports[`InstanceDetailsFooter renders per snapshot with highlightState="visible" 1`] = `
<div
  className="foot"
>
  <div
    className="highlightDiv"
  >
    <StyledIconBase
      ariaLabel="HighLight visible"
      iconName="redEye"
    />
    <StyledLabelBase>
      HighLight visible
    </StyledLabelBase>
  </div>
  <CardKebabMenuButton
    deps={
      Object {
        "cardInteractionSupport": Object {
          "supportsCopyFailureDetails": true,
          "supportsHighlighting": true,
          "supportsIssueFiling": true,
        },
      }
    }
    issueDetailsData={
      Object {
        "element": Object {
          "conciseName": "conciseName",
          "identifier": "identifier",
        },
        "howToFixSummary": "howToFixSummary",
        "rule": Object {
          "description": "description",
          "guidance": Array [
            Object {
              "href": "www.test.com",
              "tags": Array [
                Object {
                  "displayText": "New for WCAG 2.1",
                  "id": "WCAG_2_1",
                },
              ],
              "text": "text",
            },
          ],
          "id": "id",
          "url": "url",
        },
        "snippet": "snippet",
        "targetApp": Object {
          "name": "name",
          "url": "url",
        },
      }
    }
  />
</div>
`;
